src/codemagic/apple/app_store_connect/provisioning/profiles.py

```diff
@@ -131,38 +133,30 @@ def list_certificates(self, profile: Union[Profile, ResourceId]) -> List[Signing
         """
         https://developer.apple.com/documentation/appstoreconnectapi/list_all_certificates_in_a_profile
         """
-        if isinstance(profile, Profile):
-            url = profile.relationships.profiles.links.related
-        else:
-            url = f'{self.client.API_URL}/profiles/{profile}/certificates'
+        url = f'{self.client.API_URL}/profiles/{profile}/certificates'
         return [SigningCertificate(certificate) for certificate in self.client.paginate(url)]
 
     def list_certificate_ids(self, profile: Union[Profile, ResourceId]) -> List[LinkedResourceData]:
         """
         https://developer.apple.com/documentation/appstoreconnectapi/get_all_certificate_ids_in_a_profile
         """
-        if isinstance(profile, Profile):
+        url = None
+        if isinstance(profile, Profile) and profile.relationships is not None:
             url = profile.relationships.certificates.links.self
-        else:
+        if url is None:
             url = f'{self.client.API_URL}/profiles/{profile}/relationships/certificates'
         return [LinkedResourceData(certificate) for certificate in self.client.paginate(url)]
 
-    def list_devices(self, profile: Union[Profile, ResourceId]) -> List[Device]:
+    def list_devices(self, profile: ResourceId) -> List[Device]:
```

Comment:
i remember why i changed these now, it's because `Profile.relationships` doesn't have a `profiles` attribute, so it would crash if it was ever called with a `Profile`

i can't find any code that's calling it with a `Profile`, i'll just `type:ignore` it for now i guess if that's preferable to changing the signature

Reply:
These must be copy-paste errors from the past. Here url from relationship should be taken as
```python
url = profile.relationships.devices.links.self
```
and similar fixes should be applied elsewhere too.

Reply:
Your improved type checks are already paying off by revealing bugs 👍

Reply:
alright i think i fixed them all, though you probably want to double check as i'm not too sure how these methods work

Reply:
All those look good now.